### PR TITLE
feat(telemetry): Instrumented dynamic proxy

### PR DESCRIPTION
**Pull request by robzienert**

### CODE REVIEW

File: kork-telemetry/src/main/java/com/netflix/spinnaker/kork/telemetry/InstrumentedProxy.java [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+              invocationId(method)
+            ));
+          }
+          processed = true;
+          break;
+        }
+      }
+
+      if (!processed) {
+        addInstrumentedMethod(methodMetricIds, method, new MetricIds(
```
Comment (cfieber) [inline kork-telemetry/src/main/java/com/netflix/spinnaker/kork/telemetry/InstrumentedProxy.java]: feel like we either want to name the metrics after the proxied object type, and use a tag for the method name, or keep current naming and add the proxied object type as a tag

File: kork-telemetry/src/main/java/com/netflix/spinnaker/kork/telemetry/InstrumentedProxy.java [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+  private final static String INVOCATIONS = "invocations";
+  private final static String TIMING = "timing";
+
+  private final Registry registry;
+  private final Object target;
+  private final String metricNamespace;
+
+  private final Map<Method, MetricIds> instrumentedMethodMetricIds;
+
+  public InstrumentedProxy(Registry registry, Object target, String metricNamespace) {
```
Comment (cfieber) [inline kork-telemetry/src/main/java/com/netflix/spinnaker/kork/telemetry/InstrumentedProxy.java]: might allow to pass in a set of tags to add to the generated metrics here

File: kork-telemetry/src/main/java/com/netflix/spinnaker/kork/telemetry/Instrumented.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+ * limitations under the License.
+ */
+package com.netflix.spinnaker.kork.telemetry;
+
+import java.lang.annotation.*;
+
+@Documented
+@Target({ElementType.METHOD})
+@Retention(RetentionPolicy.RUNTIME)
+public @interface Instrumented {
```
Comment (cfieber) [inline kork-telemetry/src/main/java/com/netflix/spinnaker/kork/telemetry/Instrumented.java]: might allow an array of strings of tagName, tagValue to add to the metrics on here

**Comment by robzienert** [comment]
Added a test, had to do a bit of a refactor to where it now builds the metric IDs etc on-demand (caching them after first use), effectively the same functionality. Also added optional tags at both initialization-time and in the `@Instrumented` annotation.

**Pull request merged**